File: ios/NFGCrash/Games/Blocks/BlocksEngine.swift
import Foundation

enum BlocksEngine {
    static let gridSize = 8

    private static let shapes: [[[Int]]] = [
        [[0, 0]],
        [[0, 0], [0, 1]],
        [[0, 0], [0, 1], [0, 2]],
        [[0, 0], [0, 1], [0, 2], [0, 3]],
        [[0, 0], [0, 1], [1, 0], [1, 1]],
        [[0, 0], [0, 1], [1, 0]],
        [[0, 0], [0, 1], [1, 1]],
        [[0, 0], [1, 0], [1, 1]],
        [[0, 1], [1, 0], [1, 1]],
        [[0, 0], [0, 1], [0, 2], [1, 1]],
        [[0, 0], [1, 0], [1, 1], [1, 2]],
        [[0, 2], [1, 0], [1, 1], [1, 2]],
        [[0, 0], [0, 1], [1, 1], [1, 2]],
        [[0, 1], [0, 2], [1, 0], [1, 1]],
        [[0, 0], [0, 1], [0, 2], [1, 0], [1, 2]],
    ]

    static let colors = ["#22d3ee", "#a78bfa", "#4ade80", "#fb7185", "#fbbf24", "#38bdf8"]

    struct Piece: Identifiable, Equatable {
        var id: String
        var shapeId: Int
        var rotation: Int
        var colorIndex: Int
    }

    struct BoardState: Equatable {
        var grid: [[Int?]]
        var tray: [Piece?]
        var level: Int
        var linesThisLevel: Int
        var linesTarget: Int
        var totalLines: Int
        var score: Int
        var gameOver: Bool
        var rngSeed: UInt64

        static func fresh(level: Int = 1) -> BoardState {
            var state = BoardState(
                grid: Array(repeating: Array(repeating: nil, count: gridSize), count: gridSize),
                tray: [nil, nil, nil],
                level: max(1, level),
                linesThisLevel: 0,
                linesTarget: BlocksEngine.linesTarget(for: level),
                totalLines: 0,
                score: 0,
                gameOver: false,
                rngSeed: 0x4e464742_6c6b73
            )
            refillTray(&state)
            return state
        }
    }

    struct PlaceResult: Equatable {
        var clearedLines: Int
        var gameOver: Bool
    }

    static func linesTarget(for level: Int) -> Int {
        max(4, 6 + (max(1, level) - 1) * 2)
    }

    static func colorHex(index: Int) -> String {
        colors[index % colors.count]
    }

    static func pieceCells(_ piece: Piece) -> [(Int, Int)] {
        cells(shapeId: piece.shapeId, rotation: piece.rotation)
    }

    static func canPlace(_ piece: Piece, row: Int, col: Int, grid: [[Int?]]) -> Bool {
        for (dr, dc) in pieceCells(piece) {
            let r = row + dr
            let c = col + dc
            if r < 0 || c < 0 || r >= gridSize || c >= gridSize { return false }
            if grid[r][c] != nil { return false }
        }
        return true
    }

    static func hasAnyMove(_ state: BoardState) -> Bool {
        for piece in state.tray.compactMap({ $0 }) {
            for r in 0..<gridSize {
                for c in 0..<gridSize where canPlace(piece, row: r, col: c, grid: state.grid) {
                    return true
                }
            }
        }
        return false
    }

    @discardableResult
    static func placePiece(_ state: inout BoardState, piece: Piece, row: Int, col: Int) -> PlaceResult? {
        guard canPlace(piece, row: row, col: col, grid: state.grid) else { return nil }
        for (dr, dc) in pieceCells(piece) {
            state.grid[row + dr][col + dc] = piece.colorIndex
        }
        if let idx = state.tray.firstIndex(where: { $0?.id == piece.id }) {
            state.tray[idx] = nil
        }
        let cleared = clearLines(&state)
        state.linesThisLevel += cleared
        state.totalLines += cleared
        state.score += cleared * 10 * state.level
        refillTray(&state)
        if !hasAnyMove(state) { state.gameOver = true }
        return PlaceResult(clearedLines: cleared, gameOver: state.gameOver)
    }

    private static func cells(shapeId: Int, rotation: Int) -> [(Int, Int)] {
        let base = shapes[max(0, min(shapes.count - 1, shapeId))]
        var out = base.map { ($0[0], $0[1]) }
        let turns = ((rotation % 4) + 4) % 4
        for _ in 0..<turns {
            out = out.map { (r, c) in (c, -r) }
            let minR = out.map(\.0).min() ?? 0
            let minC = out.map(\.1).min() ?? 0
            out = out.map { (r, c) in (r - minR, c - minC) }
        }
        return out
    }

    private static func clearLines(_ state: inout BoardState) -> Int {
        var rows = Set<Int>()
        var cols = Set<Int>()
        var clearedCells = Set<String>()
        for r in 0..<gridSize where state.grid[r].allSatisfy({ $0 != nil }) {
            rows.insert(r)
        }
        for c in 0..<gridSize {
            var full = true
            for r in 0..<gridSize where state.grid[r][c] == nil {
                full = false
                break
            }
            if full { cols.insert(c) }
        }
        for r in rows {
            for c in 0..<gridSize {
                clearedCells.insert("\(r),\(c)")
                state.grid[r][c] = nil
            }
        }
        for c in cols {
            for r in 0..<gridSize {
                clearedCells.insert("\(r),\(c)")
                state.grid[r][c] = nil
            }
        }
        return rows.count + cols.count
    }

    private static func refillTray(_ state: inout BoardState) {
        guard state.tray.allSatisfy({ $0 == nil }) else { return }
        for i in 0..<3 {
            var piece = randomPiece(seed: &state.rngSeed)
            state.tray[i] = piece
        }
    }

    private static func randomPiece(seed: inout UInt64) -> Piece {
        var rng = seed
        let shapeId = Int(nextRandom(&rng) % UInt64(shapes.count))
        let rotation = Int(nextRandom(&rng) % 4)
        let colorIndex = Int(nextRandom(&rng) % UInt64(colors.count))
        seed = rng
        return Piece(
            id: UUID().uuidString,
            shapeId: shapeId,
            rotation: rotation,
            colorIndex: colorIndex
        )
    }

    private static func nextRandom(_ rng: inout UInt64) -> UInt64 {
        rng = rng &* 6_364_136_223_846_793_005 &+ 1
        return rng >> 32
    }
}

File: public/play/blocks-engine.js
const SHAPES = [
  [[0, 0]],
  [[0, 0], [0, 1]],
  [[0, 0], [0, 1], [0, 2]],
  [[0, 0], [0, 1], [0, 2], [0, 3]],
  [[0, 0], [0, 1], [1, 0], [1, 1]],
  [[0, 0], [0, 1], [1, 0]],
  [[0, 0], [0, 1], [1, 1]],
  [[0, 0], [1, 0], [1, 1]],
  [[0, 1], [1, 0], [1, 1]],
  [[0, 0], [0, 1], [0, 2], [1, 1]],
  [[0, 0], [1, 0], [1, 1], [1, 2]],
  [[0, 2], [1, 0], [1, 1], [1, 2]],
  [[0, 0], [0, 1], [1, 1], [1, 2]],
  [[0, 1], [0, 2], [1, 0], [1, 1]],
  [[0, 0], [0, 1], [0, 2], [1, 0], [1, 2]],
];

const COLORS = ["#22d3ee", "#a78bfa", "#4ade80", "#fb7185", "#fbbf24", "#38bdf8"];
export const GRID_SIZE = 8;

function cells(shapeId, rotation) {
  const base = SHAPES[Math.max(0, Math.min(SHAPES.length - 1, shapeId))];
  let cells = base.map(([r, c]) => [r, c]);
  const turns = ((rotation % 4) + 4) % 4;
  for (let t = 0; t < turns; t += 1) {
    cells = cells.map(([r, c]) => [c, -r]);
    const minR = Math.min(...cells.map((x) => x[0]));
    const minC = Math.min(...cells.map((x) => x[1]));
    cells = cells.map(([r, c]) => [r - minR, c - minC]);
  }
  return cells;
}

function nextRandom(rng) {
  rng = (rng * 6364136223846793005n + 1n) & 0xffffffffffffffffn;
  return { rng, val: Number(rng >> 32n) };
}

function randomPiece(rng) {
  let r = { rng: BigInt(rng || 0x4e4647426c6b73) };
  let v;
  ({ rng: r.rng, val: v } = nextRandom(r.rng));
  const shapeId = v % SHAPES.length;
  ({ rng: r.rng, val: v } = nextRandom(r.rng));
  const rotation = v % 4;
  ({ rng: r.rng, val: v } = nextRandom(r.rng));
  const colorIndex = v % COLORS.length;
  return {
    id: crypto.randomUUID(),
    shapeId,
    rotation,
    colorIndex,
    rngSeed: Number(r.rng),
  };
}

export function linesTarget(level) {
  return Math.max(4, 6 + (Math.max(1, level) - 1) * 2);
}

export function freshState(level = 1) {
  const state = {
    grid: Array.from({ length: GRID_SIZE }, () => Array(GRID_SIZE).fill(null)),
    tray: [null, null, null],
    level: Math.max(1, level),
    linesThisLevel: 0,
    linesTarget: linesTarget(level),
    totalLines: 0,
    score: 0,
    gameOver: false,
    rngSeed: 0x4e4647426c6b73,
  };
  refillTray(state);
  return state;
}

export function refillTray(state) {
  if (!state.tray.every((p) => p == null)) return;
  for (let i = 0; i < 3; i += 1) {
    const p = randomPiece(state.rngSeed);
    state.rngSeed = p.rngSeed;
    delete p.rngSeed;
    state.tray[i] = p;
  }
}

export function canPlace(piece, row, col, grid) {
  for (const [dr, dc] of cells(piece.shapeId, piece.rotation)) {
    const r = row + dr;
    const c = col + dc;
    if (r < 0 || c < 0 || r >= GRID_SIZE || c >= GRID_SIZE) return false;
    if (grid[r][c] != null) return false;
  }
  return true;
}

function clearLines(state) {
  const rows = new Set();
  const cols = new Set();
  const clearedCells = [];
  for (let r = 0; r < GRID_SIZE; r += 1) {
    if (state.grid[r].every((v) => v != null)) rows.add(r);
  }
  for (let c = 0; c < GRID_SIZE; c += 1) {
    let full = true;
    for (let r = 0; r < GRID_SIZE; r += 1) {
      if (state.grid[r][c] == null) {
        full = false;
        break;
      }
    }
    if (full) cols.add(c);
  }
  for (const r of rows) {
    for (let c = 0; c < GRID_SIZE; c += 1) {
      clearedCells.push([r, c]);
      state.grid[r][c] = null;
    }
  }
  for (const c of cols) {
    for (let r = 0; r < GRID_SIZE; r += 1) {
      if (!clearedCells.some(([rr, cc]) => rr === r && cc === c)) clearedCells.push([r, c]);
      state.grid[r][c] = null;
    }
  }
  return rows.size + cols.size;
}

export function hasAnyMove(state) {
  for (const piece of state.tray.filter(Boolean)) {
    for (let r = 0; r < GRID_SIZE; r += 1) {
      for (let c = 0; c < GRID_SIZE; c += 1) {
        if (canPlace(piece, r, c, state.grid)) return true;
      }
    }
  }
  return false;
}

export function placePiece(state, piece, row, col) {
  if (!canPlace(piece, row, col, state.grid)) return null;
  for (const [dr, dc] of cells(piece.shapeId, piece.rotation)) {
    state.grid[row + dr][col + dc] = piece.colorIndex;
  }
  const idx = state.tray.findIndex((p) => p && p.id === piece.id);
  if (idx >= 0) state.tray[idx] = null;
  const cleared = clearLines(state);
  state.linesThisLevel += cleared;
  state.totalLines += cleared;
  state.score += cleared * 10 * state.level;
  refillTray(state);
  if (!hasAnyMove(state)) state.gameOver = true;
  return { clearedLines: cleared, gameOver: state.gameOver };
}

export function pieceCells(piece) {
  return cells(piece.shapeId, piece.rotation);
}

export function colorFor(index) {
  return COLORS[index % COLORS.length];
}

export function blocksStorageKey(userId) {
  const u = String(userId || "guest").trim().toLowerCase() || "guest";
  return `nfg_blocks_web_v1_${u}`;
}

export function saveBlocksState(userId, state) {
  if (state.gameOver || !state.tray.some(Boolean)) {
    localStorage.removeItem(blocksStorageKey(userId));
    return;
  }
  localStorage.setItem(blocksStorageKey(userId), JSON.stringify(state));
}

export function loadBlocksState(userId) {
  try {
    const raw = localStorage.getItem(blocksStorageKey(userId));
    if (!raw) return null;
    const s = JSON.parse(raw);
    if (!s || s.gameOver) return null;
    return s;
  } catch {
    return null;
  }
}
